12-offer-encoding.md

```diff
@@ -0,0 +1,799 @@
+    2. data:
+        * [`tu16`:`min_final_cltv_expiry`]
+    1. type: 48 (`fallbacks`)
+    2. data:
+        * [`...*fallback_address`:`fallbacks`]
+    1. type: 50 (`payer_info`)
+    2. data:
+        * [`...*byte`:`blob`]
+    1. type: 52 (`refund_signature`)
+    2. data:
+        * [`bip340sig`:`payer_signature`]
+    1. type: 54 (`send_invoice`)
+    1. type: 56 (`replace_invoice`)
+    2. data:
+        * [`sha256`:`payment_hash`]
+    1. type: 57 (`desthint`)
+    2. data:
+        * [`point`:`source`]
+        * [`short_channel_id`:`scid`]
+        * [`u32`:`fee_base_msat`]
+        * [`u32`:`fee_proportional_millionths`]
+        * [`u16`:`cltv_expiry_delta`]
+    1. type: 240 (`signature`)
+    2. data:
+        * [`bip340sig`:`sig`]
+
+1. subtype: `blinded_payinfo`
+2. data:
+   * [`u32`:`fee_base_msat`]
+   * [`u32`:`fee_proportional_millionths`]
+   * [`u16`:`cltv_expiry_delta`]
+   * [`u16`:`flen`]
+   * [`flen*byte`:`features`]
+
+1. subtype: `fallback_address`
+2. data:
+   * [`byte`:`version`]
+   * [`u16`:`len`]
+   * [`len*byte`:`address`]
+
+## Requirements
+
+A writer of an invoice:
+  - MUST set `created_at` to the number of seconds since Midnight 1
+    January 1970, UTC when the offer was created.
+  - MUST set `payment_hash` to the SHA2 256-bit hash of the
+    `payment_preimage` that will be given in return for payment.
+  - MUST set (or not set) `send_invoice` the same as the offer.
+  - MUST set `offer_id` to the id of the offer.
+  - MUST specify exactly one signature TLV element: `signature`.
+    - MUST set `sig` to the signature using `node_id` as described in [Signature Calculation](#signature-calculation).
+  - if the chain for the invoice is not bitcoin:
+    - MUST specify `chain` the invoice is valid for.
+  - otherwise:
+    - the bitcoin chain is implied as the first and only entry.
+  - if it has bolt11 features:
+    - MUST set `features` to the bitmap of features.
+  - if the expiry for accepting payment is not 7200 seconds after `created_at`:
+    - MUST set `relative_expiry` `seconds_from_creation` to the number of
+      seconds after `created_at` that payment of this invoice should not be attempted.
+  - if the `min_final_cltv_expiry` for the last HTLC in the route is not 18:
+    - MUST set `min_final_cltv_expiry`.
+  - if it accepts onchain payments:
+    - MAY specify `fallbacks`
+    - MUST specify `fallbacks` in order of most-preferred to least-preferred
+      if it has a preference.
+    - for the bitcoin chain, it MUST set each `fallback_address` with
+      `version` as a valid witness version and `address` as a valid witness
+      program
+  - if it is connected only by private channels:
+    - if no node connecting a private channel supports blinded payments:
+      - MUST include `desthint` containing the private channel information.
+    - otherwise:
+      - MUST include `blinded_path` containing one or more paths to the node.
+  - otherwise:
+    - MAY include `blinded_path`.
```

Comment:
I think we should make `blinded_path` mandatory. The recipient can always create a single node blinded route that has the added benefit of having an `encrypted_recipient_data` compared to just using a public node id. If he invoice only has a public node id, there is no way to provide `payment_secret` or `payment_metadata` which are valuable features.

`desthint` just adds complexity for a case that will almost never happen in practice: the recipient is a private node, it wants to use offers and it is connected to a public node that support route blinding for onion messages (otherwise the recipient is not reachable by onion message and can't use offers anyways) but not route blinding for payments.

Reply:
Yeah, I agree about killing desthint.  I'll leave it in spec for now, but if it never actually gets implemented we can remove it.
@TheBlueMatt asked for it.

Hmm, we could make blinded_path mandatory in invoices (where space doesn't matter), but it does introduce requirements at the payment layer which we've avoided so far.  You can generate a bolt11 from a bolt12 invoice for the simple cases for now, for example.

At the moment my implementation uses the merkle of the invoice as the `payment_secret`, but ideally this would be the signature of that using the `payer_key`: invoices were never supposed to be secret, and it'd be far better to use the payer_key as a real form of anti-probing.  I could be swayed, however...